File: CRUD/PHP/Ajax V3/includes/AccesoDatos.php
<?php

// Written by Ismael Heredia in the year 2017

include_once("Conexion.php");

class AccesoDatos {
  
   public function __construct(){

   }

   // Productos

   public function listarProductos($patron) {
      $productos = array();
   	  $conexion = new Conexion();
      $conexion->abrir_conexion();
      $conn = $conexion->retornar_conexion();
      $sql = $conn->prepare('SELECT * FROM productos WHERE nombre LIKE :patron');
      $sql->execute(array('patron' => '%'.$patron.'%'));
      $resultado = $sql->fetchAll();
      foreach ($resultado as $fila) {
        $id = $fila['id'];
        $nombre = $fila['nombre'];
        $descripcion = $fila['descripcion'];
        $precio = $fila['precio'];
        $id_proveedor = $fila['id_proveedor'];
        $fecha_registro = $fila['fecha_registro'];
        $producto = Producto::CreateProducto($id,$nombre,$descripcion,$precio,$id_proveedor,$fecha_registro);
        array_push($productos,$producto);
      }
      foreach ($productos as $producto) {
        $sql = $conn->prepare('SELECT * FROM proveedores WHERE id = :id');
        $sql->execute(array('id' => $producto->getId_proveedor()));
        $resultado = $sql->fetch();  
        $id = $resultado['id'];
        $nombre = $resultado['nombre'];
        $direccion = $resultado['direccion'];
        $telefono = $resultado['telefono'];
        $fecha_registro = $resultado['fecha_registro']; 
        $proveedor = Proveedor::CreateProveedor($id,$nombre,$direccion,$telefono,$fecha_registro);
        $producto->setProveedor($proveedor);     
      }
      $conexion->cerrar_conexion();
      return $productos;   
   }

   public function cargarProducto($id) {
   	  $conexion = new Conexion();
      $conexion->abrir_conexion();
      $conn = $conexion->retornar_conexion();
      $sql = $conn->prepare('SELECT * FROM productos WHERE id = :id');
      $sql->execute(array('id' => $id));
      $resultado = $sql->fetch();
      $id = $resultado['id'];
      $nombre = $resultado['nombre'];
      $descripcion = $resultado['descripcion'];
      $precio = $resultado['precio'];
      $id_proveedor = $resultado['id_proveedor'];
      $fecha_registro = $resultado['fecha_registro'];
      $producto = Producto::CreateProducto($id,$nombre,$descripcion,$precio,$id_proveedor,$fecha_registro);
      $sql = $conn->prepare('SELECT * FROM proveedores WHERE id = :id');
      $sql->execute(array('id' => $producto->getId_proveedor()));
      $resultado = $sql->fetch();  
      $id = $resultado['id'];
      $nombre = $resultado['nombre'];
      $direccion = $resultado['direccion'];
      $telefono = $resultado['telefono'];
      $fecha_registro = $resultado['fecha_registro']; 
      $proveedor = Proveedor::CreateProveedor($id,$nombre,$direccion,$telefono,$fecha_registro);
      $producto->setProveedor($proveedor);  
      $conexion->cerrar_conexion();
      return $producto;   
   }  

   public function agregarProducto($producto) {
      $nombre = $producto->getNombre();
      $descripcion = $producto->getDescripcion();
      $precio = $producto->getPrecio();
      $id_proveedor = $producto->getId_proveedor();
      $fecha_registro = $producto->getFecha_registro();
      try {
        $conexion = new Conexion();
        $conexion->abrir_conexion();
        $conn = $conexion->retornar_conexion();
        $sql = $conn->prepare('INSERT INTO productos(nombre,descripcion,precio,id_proveedor,fecha_registro) VALUES(:nombre,:descripcion,:precio,:id_proveedor,:fecha_registro)');
        $sql->execute(array('nombre' => $nombre,'descripcion' => $descripcion,'precio' => $precio,'id_proveedor' => $id_proveedor,'fecha_registro'=>$fecha_registro)); 
        $conexion->cerrar_conexion();
        return true;
      } catch (PDOException $e) {
        return false;
      }     
   }

   public function editarProducto($producto) {
      $id = $producto->getId();
      $nombre = $producto->getNombre();
      $descripcion = $producto->getDescripcion();
      $precio = $producto->getPrecio();
      $id_proveedor = $producto->getId_proveedor();
      $fecha_registro = $producto->getFecha_registro();
      try {
        $conexion = new Conexion();
        $conexion->abrir_conexion();
        $conn = $conexion->retornar_conexion();
        $sql = $conn->prepare('UPDATE productos SET nombre = :nombre, descripcion = :descripcion, precio = :precio, id_proveedor = :id_proveedor WHERE id = :id');
        $sql->execute(array('nombre' => $nombre,'descripcion' => $descripcion,'precio' => $precio,'id_proveedor' => $id_proveedor,'id'=>$id)); 
        $conexion->cerrar_conexion();
        return true;
      } catch (PDOException $e) {
        return false;
      } 
   }

   public function borrarProducto($producto) {
      $id = $producto->getId();
      try {
        $conexion = new Conexion();
        $conexion->abrir_conexion();
        $conn = $conexion->retornar_conexion();
        $sql = $conn->prepare('DELETE FROM productos WHERE id = :id');
        $sql->execute(array('id'=>$id)); 
        $conexion->cerrar_conexion();
        return true;
      } catch (PDOException $e) {
        return false;
      } 
   }

   public function comprobarExistenciaProductoCrear($nombre) {
      $response = false;
      $conexion = new Conexion();
      $conexion->abrir_conexion();
      $conn = $conexion->retornar_conexion();
      $sql = $conn->prepare('SELECT * FROM productos WHERE nombre = :nombre');
      $sql->execute(array('nombre' => $nombre));
      $resultado = $sql->fetchAll();
      $cantidad = count($resultado);
      if($cantidad >= 1) {
        $response = true;
      } else {
        $response = false;
      }
      $conexion->cerrar_conexion();
      return $response;
   }

   public function comprobarExistenciaProductoEditar($id,$nombre) {
      $response = false;
      $conexion = new Conexion();
      $conexion->abrir_conexion();
      $conn = $conexion->retornar_conexion();
      $sql = $conn->prepare('SELECT * FROM productos WHERE nombre = :nombre AND id != :id');
      $sql->execute(array('nombre' => $nombre,'id' => $id));
      $resultado = $sql->fetchAll();
      $cantidad = count($resultado);
      if($cantidad >= 1) {
        $response = true;
      } else {
        $response = false;
      }
      $conexion->cerrar_conexion();
      return $response;
   }

   // Proveedores

   public function listarProveedores($patron) {
      $proveedores = array();
   	  $conexion = new Conexion();
      $conexion->abrir_conexion();
      $conn = $conexion->retornar_conexion();
      $sql = $conn->prepare('SELECT * FROM proveedores WHERE nombre LIKE :patron');
      $sql->execute(array('patron' => '%'.$patron.'%'));
      $resultado = $sql->fetchAll();
      foreach ($resultado as $fila) {
        $id = $fila['id'];
        $nombre = $fila['nombre'];
        $direccion = $fila['direccion'];
        $telefono = $fila['telefono'];
        $fecha_registro = $fila['fecha_registro'];
        $proveedor = Proveedor::CreateProveedor($id,$nombre,$direccion,$telefono,$fecha_registro);
        array_push($proveedores,$proveedor);
      }
      $conexion->cerrar_conexion();
      return $proveedores; 
   }

   public function cargarProveedor($id) {
   	  $conexion = new Conexion();
      $conexion->abrir_conexion();
      $conn = $conexion->retornar_conexion();
      $sql = $conn->prepare('SELECT * FROM proveedores WHERE id = :id');
      $sql->execute(array('id' => $id));
      $resultado = $sql->fetch();
      $id = $resultado['id'];
      $nombre = $resultado['nombre'];
      $direccion = $resultado['direccion'];
      $telefono = $resultado['telefono'];
      $fecha_registro = $resultado['fecha_registro'];
      $proveedor = Proveedor::CreateProveedor($id,$nombre,$direccion,$telefono,$fecha_registro);
      $conexion->cerrar_conexion();
      return $proveedor; 
   }

   public function agregarProveedor($proveedor) {
      $nombre = $proveedor->getNombre();
      $direccion = $proveedor->getDireccion();
      $telefono = $proveedor->getTelefono();
      $fecha_registro = $proveedor->getFecha_registro();
      try {
        $conexion = new Conexion();
        $conexion->abrir_conexion();
        $conn = $conexion->retornar_conexion();
        $sql = $conn->prepare('INSERT INTO proveedores(nombre,direccion,telefono,fecha_registro) VALUES(:nombre,:direccion,:telefono,:fecha_registro)');
        $sql->execute(array('nombre' => $nombre,'direccion' => $direccion,'telefono' => $telefono,'fecha_registro'=>$fecha_registro)); 
        $conexion->cerrar_conexion();
        return true;
      } catch (PDOException $e) {
        return false;
      }     
   }

   public function editarProveedor($proveedor) {
      $id = $proveedor->getId();
      $nombre = $proveedor->getNombre();
      $direccion = $proveedor->getDireccion();
      $telefono = $proveedor->getTelefono();
      $fecha_registro = $proveedor->getFecha_registro();
      try {
        $conexion = new Conexion();
        $conexion->abrir_conexion();
        $conn = $conexion->retornar_conexion();
        $sql = $conn->prepare('UPDATE proveedores SET nombre = :nombre, direccion = :direccion, telefono = :telefono WHERE id = :id');
        $sql->execute(array('nombre' => $nombre,'direccion' => $direccion,'telefono' => $telefono,'id'=>$id)); 
        $conexion->cerrar_conexion();
        return true;
      } catch (PDOException $e) {
        return false;
      } 
   }

   public function borrarProveedor($proveedor) {
      $id = $proveedor->getId();
      try {
        $conexion = new Conexion();
        $conexion->abrir_conexion();
        $conn = $conexion->retornar_conexion();
        $sql = $conn->prepare('DELETE FROM proveedores WHERE id = :id');
        $sql->execute(array('id'=>$id)); 
        $conexion->cerrar_conexion();
        return true;
      } catch (PDOException $e) {
        return false;
      } 
   }

   public function comprobarExistenciaProveedorCrear($nombre) {
      $response = false;
      $conexion = new Conexion();
      $conexion->abrir_conexion();
      $conn = $conexion->retornar_conexion();
      $sql = $conn->prepare('SELECT * FROM proveedores WHERE nombre = :nombre');
      $sql->execute(array('nombre' => $nombre));
      $resultado = $sql->fetchAll();
      $cantidad = count($resultado);
      if($cantidad >= 1) {
        $response = true;
      } else {
        $response = false;
      }
      $conexion->cerrar_conexion();
      return $response;
   }

   public function comprobarExistenciaProveedorEditar($id,$nombre) {
      $response = false;
      $conexion = new Conexion();
      $conexion->abrir_conexion();
      $conn = $conexion->retornar_conexion();
      $sql = $conn->prepare('SELECT * FROM proveedores WHERE nombre = :nombre AND id != :id');
      $sql->execute(array('nombre' => $nombre,'id' => $id));
      $resultado = $sql->fetchAll();
      $cantidad = count($resultado);
      if($cantidad >= 1) {
        $response = true;
      } else {
        $response = false;
      }
      $conexion->cerrar_conexion();
      return $response;
   }

   // Usuarios

   public function listarUsuarios($patron) {
      $usuarios = array();
   	  $conexion = new Conexion();
      $conexion->abrir_conexion();
      $conn = $conexion->retornar_conexion();
      $sql = $conn->prepare('SELECT * FROM usuarios WHERE nombre LIKE :patron');
      $sql->execute(array('patron' => '%'.$patron.'%'));
      $resultado = $sql->fetchAll();
      foreach ($resultado as $fila) {
        $id = $fila['id'];
        $nombre = $fila['nombre'];
        $clave = $fila['clave'];
        $tipo = $fila['id_tipo'];
        $fecha_registro = $fila['fecha_registro'];
        $usuario = Usuario::CreateUsuario($id,$nombre,$clave,$tipo,$fecha_registro);
        array_push($usuarios,$usuario);
      }
      foreach ($usuarios as $usuario) {
        $sql = $conn->prepare('SELECT * FROM tipos_usuarios WHERE id = :id');
        $sql->execute(array('id' => $usuario->getId_tipo()));
        $resultado = $sql->fetch();
        $id = $resultado['id'];
        $nombre = $resultado['nombre'];
        $tipo = TipoUsuario::CreateTipoUsuario($id,$nombre); 
        $usuario->setTipo($tipo);       
      }
      $conexion->cerrar_conexion();
      return $usuarios;  
   }

   public function cargarUsuario($id) {
   	  $conexion = new Conexion();
      $conexion->abrir_conexion();
      $conn = $conexion->retornar_conexion();
      $sql = $conn->prepare('SELECT * FROM usuarios WHERE id = :id');
      $sql->execute(array('id' => $id));
      $resultado = $sql->fetch();
      $id = $resultado['id'];
      $nombre = $resultado['nombre'];
      $clave = $resultado['clave'];
      $tipo = $resultado['id_tipo'];
      $fecha_registro = $resultado['fecha_registro'];
      $usuario = Usuario::CreateUsuario($id,$nombre,$clave,$tipo,$fecha_registro);
      $sql = $conn->prepare('SELECT * FROM tipos_usuarios WHERE id = :id');
      $sql->execute(array('id' => $usuario->getId_tipo()));
      $resultado = $sql->fetch();
      $id = $resultado['id'];
      $nombre = $resultado['nombre'];
      $tipo = TipoUsuario::CreateTipoUsuario($id,$nombre); 
      $usuario->setTipo($tipo);   
      $conexion->cerrar_conexion();
      return $usuario;  
   }

   public function agregarUsuario($usuario) {
      $nombre = $usuario->getNombre();
      $clave = $usuario->getClave();
      $id_tipo = $usuario->getId_tipo();
      $fecha_registro = $usuario->getFecha_registro();
      try {
        $conexion = new Conexion();
        $conexion->abrir_conexion();
        $conn = $conexion->retornar_conexion();
        $sql = $conn->prepare('INSERT INTO usuarios(nombre,clave,id_tipo,fecha_registro) VALUES(:nombre,:clave,:id_tipo,:fecha_registro)');
        $sql->execute(array('nombre' => $nombre,'clave' => $clave,'id_tipo' => $id_tipo,'fecha_registro'=>$fecha_registro)); 
        $conexion->cerrar_conexion();
        return true;
      } catch (PDOException $e) {
        return false;
      }     
   }

   public function editarUsuario($usuario) {
      $id = $usuario->getId();
      $id_tipo = $usuario->getId_tipo();
      try {
        $conexion = new Conexion();
        $conexion->abrir_conexion();
        $conn = $conexion->retornar_conexion();
        $sql = $conn->prepare('UPDATE usuarios SET id_tipo = :id_tipo WHERE id = :id');
        $sql->execute(array('id_tipo' => $id_tipo,'id'=>$id)); 
        $conexion->cerrar_conexion();
        return true;
      } catch (PDOException $e) {
        return false;
      } 
   }

   public function borrarUsuario($usuario) {
      $id = $usuario->getId();
      try {
        $conexion = new Conexion();
        $conexion->abrir_conexion();
        $conn = $conexion->retornar_conexion();
        $sql = $conn->prepare('DELETE FROM usuarios WHERE id = :id');
        $sql->execute(array('id'=>$id)); 
        $conexion->cerrar_conexion();
        return true;
      } catch (PDOException $e) {
        return false;
      } 
   }

   public function comprobarExistenciaUsuarioCrear($nombre) {
      $response = false;
      $conexion = new Conexion();
      $conexion->abrir_conexion();
      $conn = $conexion->retornar_conexion();
      $sql = $conn->prepare('SELECT * FROM usuarios WHERE nombre = :nombre');
      $sql->execute(array('nombre' => $nombre));
      $resultado = $sql->fetchAll();
      $cantidad = count($resultado);
      if($cantidad >= 1) {
        $response = true;
      } else {
        $response = false;
      }
      $conexion->cerrar_conexion();
      return $response;
   }

   public function comprobarExistenciaUsuarioEditar($id,$nombre) {
      $response = false;
      $conexion = new Conexion();
      $conexion->abrir_conexion();
      $conn = $conexion->retornar_conexion();
      $sql = $conn->prepare('SELECT * FROM usuarios WHERE nombre = :nombre AND id != :id');
      $sql->execute(array('nombre' => $nombre,'id' => $id));
      $resultado = $sql->fetchAll();
      $cantidad = count($resultado);
      if($cantidad >= 1) {
        $response = true;
      } else {
        $response = false;
      }
      $conexion->cerrar_conexion();
      return $response;
   }

   // Tipos de usuarios

   public function listarTiposUsuarios($patron) {
   	  $conexion = new Conexion();
      $conexion->abrir_conexion();
      $conn = $conexion->retornar_conexion();
      $tipos = array();
      $sql = $conn->prepare('SELECT * FROM tipos_usuarios WHERE nombre LIKE :patron');
      $sql->execute(array('patron' => '%'.$patron.'%'));
      $resultado = $sql->fetchAll();
      foreach ($resultado as $fila) {
        $id = $fila['id'];
        $nombre = $fila['nombre'];
        $tipo = TipoUsuario::CreateTipoUsuario($id,$nombre);
        array_push($tipos,$tipo);
      }
      $conexion->cerrar_conexion();
      return $tipos;  
   }

   // Controles de usuarios

   public function ingreso_usuario($usuario,$clave) {
      $response = false;
   	  $conexion = new Conexion();
      $conexion->abrir_conexion();
      $conn = $conexion->retornar_conexion();
      $sql = $conn->prepare('SELECT id FROM usuarios WHERE nombre = :usuario AND clave = :clave');
      $sql->execute(array('usuario' => $usuario, 'clave' => $clave));
      $resultado = $sql->fetchAll();
      $cantidad = count($resultado);
      if($cantidad >= 1) {
        $response = true;
      } else {
        $response = false;
      }
      $conexion->cerrar_conexion();
      return $response;
   }
   
   public function es_admin($usuario) {
      $response = false;
   	  $conexion = new Conexion();
      $conexion->abrir_conexion();
      $conn = $conexion->retornar_conexion();
      $sql = $conn->prepare('SELECT id_tipo FROM usuarios WHERE nombre = :usuario');
      $sql->execute(array('usuario' => $usuario));
      $resultado = $sql->fetch();
      $id_tipo = $resultado['id_tipo'];
      if($id_tipo=='1') {
        $response = true;
      } else {
        $response = false;
      }
      $conexion->cerrar_conexion();
      return $response;
   }

   // Cuenta

   public function cambiarUsuario($usuario,$nuevo_usuario) {
      try {
        $conexion = new Conexion();
        $conexion->abrir_conexion();
        $conn = $conexion->retornar_conexion();
        $sql = $conn->prepare('UPDATE usuarios SET nombre = :nuevo_usuario WHERE nombre = :usuario');
        $sql->execute(array('usuario'=>$usuario,'nuevo_usuario'=>$nuevo_usuario)); 
        $conexion->cerrar_conexion();
        return true;
      } catch (PDOException $e) {
        return false;
      } 
   }

   public function cambiarClave($usuario,$nueva_clave) {
      try {
        $conexion = new Conexion();
        $conexion->abrir_conexion();
        $conn = $conexion->retornar_conexion();
        $sql = $conn->prepare('UPDATE usuarios SET clave = :nueva_clave WHERE nombre = :usuario');
        $sql->execute(array('usuario'=>$usuario,'nueva_clave'=>$nueva_clave)); 
        $conexion->cerrar_conexion();
        return true;
      } catch (PDOException $e) {
        return false;
      } 
   }
            	   
   public function __destruct(){
   }  
   
}

?>

File: CRUD/Java usando NetBeans/Aplicacion Web usando Hibernate/SistemaWebHibernate/SistemaWebHibernate/src/main/java/sistema/controlador/AccesoDatos.java
// Written by Ismael Heredia in the year 2016
package sistema.controlador;

import org.hibernate.HibernateException;
import org.hibernate.Query;
import org.hibernate.Session;
import sistema.entity.Productos;
import sistema.entity.Proveedores;
import sistema.entity.Usuarios;
import sistema.util.HibernateUtil;
import java.util.ArrayList;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import java.util.ArrayList;
import org.hibernate.HibernateException;
import org.hibernate.Query;
import org.hibernate.Session;
import sistema.util.HibernateUtil;

public class AccesoDatos {

    public Integer ejecutarConsulta(String query) {
        int cantidad = 0;
        Session session = HibernateUtil.getSessionFactory().openSession();
        session.beginTransaction();
        Query q = session.createQuery(query);
        ArrayList lista = (ArrayList) q.list();
        return lista.size();
    }

    public boolean ingreso_usuario(String username, String password) {
        boolean respuesta = false;
        try {
            int cantidad = ejecutarConsulta("from Usuarios where usuario='" + username + "' and clave='" + password + "'");
            if (cantidad >= 1) {
                respuesta = true;
            } else {
                respuesta = false;
            }
        } catch (Exception e) {
            System.out.println(e);
        }
        return respuesta;
    }
    
    public boolean es_admin(String username) {
        boolean respuesta = false;
        try {
            String query = "from Usuarios where usuario='" + username + "'";
            Session session = HibernateUtil.getSessionFactory().openSession();
            session.beginTransaction();
            Query q = session.createQuery(query);
            ArrayList lista = (ArrayList) q.list();
            for (Object o : lista) {
                Usuarios usuario = (Usuarios) o;
                int tipo = usuario.getTipo();
                if (tipo == 1) {
                    respuesta = true;
                } else {
                    respuesta = false;
                }
            }
        } catch (Exception e) {
            //
        }
        return respuesta;
    }
    
    public int get_id_by_username(String username) {
        int id_usuario = 0;
        try {
            String query = "from Usuarios where usuario='" + username + "'";
            Session session = HibernateUtil.getSessionFactory().openSession();
            session.beginTransaction();
            Query q = session.createQuery(query);
            ArrayList lista = (ArrayList) q.list();
            for (Object o : lista) {
                Usuarios usuario = (Usuarios) o;
                id_usuario = usuario.getIdUsuario();
            }
        } catch (Exception e) {
            //
        }
        return id_usuario;
    }
    
    public Boolean agregarProveedor(String nombre_proveedor, String direccion, int telefono, String fecha_registro) {
        boolean response = false;
        try {
            Session session = HibernateUtil.getSessionFactory().openSession();
            session.beginTransaction();
            Proveedores proveedor = new Proveedores();
            proveedor.setNombreEmpresa(nombre_proveedor);
            proveedor.setDireccion(direccion);
            proveedor.setTelefono(telefono);
            proveedor.setFechaRegistroProveedor(fecha_registro);
            session.save(proveedor);
            session.getTransaction().commit();
            response = true;
        } catch (HibernateException he) {
            response = false;
        }
        return response;
    }
    
    public Boolean editarProveedor(int id_proveedor, String nombre_proveedor, String direccion, int telefono, String fecha_registro) {
        boolean response = false;
        try {
            Session session = HibernateUtil.getSessionFactory().openSession();
            session.beginTransaction();
            Proveedores proveedor = (Proveedores) session.get(Proveedores.class, id_proveedor);
            proveedor.setNombreEmpresa(nombre_proveedor);
            proveedor.setDireccion(direccion);
            proveedor.setTelefono(telefono);
            if (!"".equals(fecha_registro)) {
                proveedor.setFechaRegistroProveedor(fecha_registro);
            }
            session.update(proveedor);
            session.getTransaction().commit();
            response = true;
        } catch (HibernateException he) {
            response = false;
        }
        return response;
    }
    
    public Boolean borrarProveedor(int id_proveedor) {
        boolean response = false;
        try {
            Session session = HibernateUtil.getSessionFactory().openSession();
            session.beginTransaction();
            Proveedores proveedor = new Proveedores();
            proveedor.setIdProveedor(id_proveedor);
            session.delete(proveedor);
            session.getTransaction().commit();
            response = true;
        } catch (HibernateException he) {
            response = false;
        }
        return response;
    }
    
    public Boolean agregarProducto(String nombre_producto, String descripcion, int precio, int id_proveedor, String fecha_registro) {
        boolean response = false;
        try {
            Session session = HibernateUtil.getSessionFactory().openSession();
            session.beginTransaction();
            Productos producto = new Productos();
            producto.setNombreProducto(nombre_producto);
            producto.setDescripcion(descripcion);
            producto.setPrecio(precio);
            Proveedores proveedor = (Proveedores) session.get(Proveedores.class, id_proveedor);
            producto.setProveedores(proveedor);
            producto.setFechaRegistro(fecha_registro);
            session.save(producto);
            session.getTransaction().commit();
            response = true;
        } catch (HibernateException he) {
            response = false;
        }
        return response;
    }
    
    public Boolean editarProducto(int id_producto, String nombre_producto, String descripcion, int precio, int id_proveedor, String fecha_registro) {
        boolean response = false;
        try {
            Session session = HibernateUtil.getSessionFactory().openSession();
            session.beginTransaction();
            Productos producto = (Productos) session.get(Productos.class, id_producto);
            producto.setNombreProducto(nombre_producto);
            producto.setDescripcion(descripcion);
            producto.setPrecio(precio);
            Proveedores proveedor = (Proveedores) session.get(Proveedores.class, id_proveedor);
            producto.setProveedores(proveedor);
            if (!"".equals(fecha_registro)) {
                producto.setFechaRegistro(fecha_registro);
            }
            session.update(producto);
            session.getTransaction().commit();
            response = true;
        } catch (HibernateException he) {
            response = false;
        }
        return response;
    }
    
    public Boolean borrarProducto(int id_producto) {
        boolean response = false;
        try {
            Session session = HibernateUtil.getSessionFactory().openSession();
            session.beginTransaction();
            Productos producto = new Productos();
            producto.setIdProducto(id_producto);
            session.delete(producto);
            session.getTransaction().commit();
            response = true;
        } catch (HibernateException he) {
            response = false;
        }
        return response;
    }
    
    public Boolean agregarUsuario(String nombre_usuario, String clave, int tipo, String fecha_registro) {
        boolean response = false;
        try {
            Session session = HibernateUtil.getSessionFactory().openSession();
            session.beginTransaction();
            Usuarios usuario = new Usuarios();
            usuario.setUsuario(nombre_usuario);
            usuario.setClave(clave);
            usuario.setTipo(tipo);
            usuario.setFechaRegistro(fecha_registro);
            session.save(usuario);
            session.getTransaction().commit();
            response = true;
        } catch (HibernateException he) {
            response = false;
        }
        return response;
    }
    
    public Boolean editarUsuario(int id_usuario, String nombre_usuario, String clave, int tipo, String fecha_registro) {
        boolean response = false;
        try {
            Session session = HibernateUtil.getSessionFactory().openSession();
            session.beginTransaction();
            Usuarios usuario = (Usuarios) session.get(Usuarios.class, id_usuario);
            if (!"".equals(nombre_usuario)) {
                usuario.setUsuario(nombre_usuario);
            }
            if (!"".equals(clave)) {
                usuario.setClave(clave);
            }
            if (tipo > 0) {
                usuario.setTipo(tipo);
            }
            if (!"".equals(fecha_registro)) {
                usuario.setFechaRegistro(fecha_registro);
            }
            session.update(usuario);
            session.getTransaction().commit();
            response = true;
        } catch (HibernateException he) {
            response = false;
        }
        return response;
    }
    
    public Boolean borrarUsuario(int id_usuario) {
        boolean response = false;
        try {
            Session session = HibernateUtil.getSessionFactory().openSession();
            session.beginTransaction();
            Usuarios usuario = new Usuarios();
            usuario.setIdUsuario(id_usuario);
            session.delete(usuario);
            session.getTransaction().commit();
            response = true;
        } catch (HibernateException he) {
            response = false;
        }
        return response;
    }
    
    public boolean comprobar_existencia_producto_crear(String nombre_producto) {
        boolean respuesta = false;
        Session session = HibernateUtil.getSessionFactory().openSession();
        session.beginTransaction();
        Query q = session.createQuery("from Productos p where p.nombreProducto like :patron");
        q.setString("patron", "%" + nombre_producto + "%");
        ArrayList lista = (ArrayList) q.list();
        if (lista.size() >= 1) {
            respuesta = true;
        } else {
            respuesta = false;
        }
        return respuesta;
    }
    
    public boolean comprobar_existencia_producto_editar(int id_producto, String nombre_producto) {
        boolean respuesta = false;
        Session session = HibernateUtil.getSessionFactory().openSession();
        session.beginTransaction();
        Query q = session.createQuery("from Productos p where p.nombreProducto like :patron and p.idProducto!=" + id_producto);
        q.setString("patron", "%" + nombre_producto + "%");
        ArrayList lista = (ArrayList) q.list();
        if (lista.size() >= 1) {
            respuesta = true;
        } else {
            respuesta = false;
        }
        return respuesta;
    }
    
    public boolean comprobar_existencia_proveedor_crear(String nombre_empresa) {
        boolean respuesta = false;
        Session session = HibernateUtil.getSessionFactory().openSession();
        session.beginTransaction();
        Query q = session.createQuery("from Proveedores p where p.nombreEmpresa like :patron");
        q.setString("patron", "%" + nombre_empresa + "%");
        ArrayList lista = (ArrayList) q.list();
        if (lista.size() >= 1) {
            respuesta = true;
        } else {
            respuesta = false;
        }
        return respuesta;
    }
    
    public boolean comprobar_existencia_proveedor_editar(int id_proveedor, String nombre_empresa) {
        boolean respuesta = false;
        Session session = HibernateUtil.getSessionFactory().openSession();
        session.beginTransaction();
        Query q = session.createQuery("from Proveedores p where p.nombreEmpresa like :patron and p.idProveedor!=" + id_proveedor);
        q.setString("patron", "%" + nombre_empresa + "%");
        ArrayList lista = (ArrayList) q.list();
        if (lista.size() >= 1) {
            respuesta = true;
        } else {
            respuesta = false;
        }
        return respuesta;
    }
    
    public boolean comprobar_existencia_usuario_crear(String nombre_usuario) {
        boolean respuesta = false;
        Session session = HibernateUtil.getSessionFactory().openSession();
        session.beginTransaction();
        Query q = session.createQuery("from Usuarios u where u.usuario like :patron");
        q.setString("patron", "%" + nombre_usuario + "%");
        ArrayList lista = (ArrayList) q.list();
        if (lista.size() >= 1) {
            respuesta = true;
        } else {
            respuesta = false;
        }
        return respuesta;
    }
    
    public boolean comprobar_existencia_usuario_editar(String nombre_usuario) {
        boolean respuesta = false;
        Session session = HibernateUtil.getSessionFactory().openSession();
        session.beginTransaction();
        Query q = session.createQuery("from Usuarios u where u.usuario like :patron");
        q.setString("patron", "%" + nombre_usuario + "%");
        ArrayList lista = (ArrayList) q.list();
        if (lista.size() >= 1) {
            respuesta = true;
        } else {
            respuesta = false;
        }
        return respuesta;
    }
    
    public ArrayList listarProveedoresEnTabla(String patron) {
        ArrayList listaProveedores = new ArrayList();
        try {
            Session session = HibernateUtil.getSessionFactory().openSession();
            session.beginTransaction();
            Query q = session.createQuery("from Proveedores p where p.nombreEmpresa like :patron");
            q.setString("patron", "%" + patron + "%");
            ArrayList lista = (ArrayList) q.list();
            for (Object o : lista) {
                Proveedores proveedor = (Proveedores) o;
                listaProveedores.add(proveedor);
            }
        } catch (Exception e) {
            //
        }
        return listaProveedores;
    }
    
    public Proveedores cargarProveedorEnTabla(int id_proveedor_to_load) {
        Proveedores proveedor = new Proveedores();
        try {
            Session session = HibernateUtil.getSessionFactory().openSession();
            session.beginTransaction();
            proveedor = (Proveedores) session.get(Proveedores.class, id_proveedor_to_load);
            session.getTransaction().commit();
        } catch (HibernateException he) {
            //
        }
        return proveedor;
    }
    
    public ArrayList listarProductosEnTabla(String patron) {
        ArrayList listaProductos = new ArrayList();
        try {
            Session session = HibernateUtil.getSessionFactory().openSession();
            session.beginTransaction();
            Query q = session.createQuery("from Productos p where p.nombreProducto like :patron");
            q.setString("patron", "%" + patron + "%");
            ArrayList lista = (ArrayList) q.list();
            for (Object o : lista) {
                Productos producto = (Productos) o;
                listaProductos.add(producto);
            }
        } catch (Exception e) {
            //
        }
        return listaProductos;
    }
    
    public Productos cargarProductoEnTabla(int id_producto_to_load) {
        Productos producto = new Productos();
        try {
            Session session = HibernateUtil.getSessionFactory().openSession();
            session.beginTransaction();
            producto = (Productos) session.get(Productos.class, id_producto_to_load);
            session.getTransaction().commit();
        } catch (HibernateException he) {
            //
        }
        return producto;
    }
    
    public ArrayList listarUsuariosEnTabla(String patron) {
        ArrayList listaUsuarios = new ArrayList();
        try {
            Session session = HibernateUtil.getSessionFactory().openSession();
            session.beginTransaction();
            Query q = session.createQuery("from Usuarios u where u.usuario like :patron");
            q.setString("patron", "%" + patron + "%");
            ArrayList lista = (ArrayList) q.list();
            for (Object o : lista) {
                Usuarios usuario = (Usuarios) o;
                listaUsuarios.add(usuario);
            }
        } catch (Exception e) {
            //
        }
        return listaUsuarios;
    }
    
    public Usuarios cargarUsuarioEnTabla(int id_usuario_to_load) {
        Usuarios usuario = new Usuarios();
        try {
            Session session = HibernateUtil.getSessionFactory().openSession();
            session.beginTransaction();
            usuario = (Usuarios) session.get(Usuarios.class, id_usuario_to_load);
            session.getTransaction().commit();
        } catch (HibernateException he) {
            //
        }
        return usuario;
    }
    
    public String cargarNombreProveedor(int id_proveedor_to_load) {
        String nombre_empresa = "";
        try {
            Session session = HibernateUtil.getSessionFactory().openSession();
            session.beginTransaction();
            Proveedores proveedor = (Proveedores) session.get(Proveedores.class, id_proveedor_to_load);
            nombre_empresa = proveedor.getNombreEmpresa();
            session.getTransaction().commit();
        } catch (HibernateException he) {
            //
        }
        return nombre_empresa;
    }
    
}
